apiserver/application/application.go

```diff
@@ -1049,54 +1034,107 @@ func (api *API) differentControllerOfferedApplication(userName, modelName, offer
 	//return &offers[0], sourceModelTag, nil
 }
 
-func (api *API) makeOfferParamsFromApplication(offerName string, app *state.Application) (*params.ApplicationOffer, error) {
-	ch, _, err := app.Charm()
+func (api *API) makeOfferParams(st *state.State, offer *jujucrossmodel.ApplicationOffer) (
+	*params.ApplicationOffer, error,
+) {
+	app, err := st.Application(offer.ApplicationName)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	appBindings, err := app.EndpointBindings()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	result := params.ApplicationOffer{
-		SourceModelTag:         api.backend.ModelTag().String(),
-		OfferName:              offerName,
-		ApplicationDescription: ch.Meta().Description,
+		OfferName:              offer.OfferName,
+		ApplicationDescription: offer.ApplicationDescription,
 	}
-	eps, err := app.Endpoints()
-	for _, ep := range eps {
+
+	spaceNames := set.NewStrings()
+	for _, ep := range offer.Endpoints {
 		result.Endpoints = append(result.Endpoints, params.RemoteEndpoint{
 			Name:      ep.Name,
 			Interface: ep.Interface,
 			Role:      ep.Role,
 			Scope:     ep.Scope,
 			Limit:     ep.Limit,
 		})
+		spaceNames.Add(appBindings[ep.Name])
```

Comment:
My thinking was that if it didn't contain the endpoint then the result would be "" anyway, which is what we'd want here, because we want to get the provider info for the default space in that case. But I don't want to error if there's no binding. (I'll do a check to confirm that there isn't always an entry for an endpoint if it hasn't been explicitly bound.)